external/llvm-project/mlir/lib/Conversion/AMDGPUToROCDL/AMDGPUToROCDL.cpp

```diff
@@ -830,6 +830,151 @@ LogicalResult PackedStochRoundFp8OpLowering::matchAndRewrite(
+    }
+
+    // This is taken from the following file llvm/lib/Target/AMDGPU/SIDefines.h
+    enum DppCtrl : unsigned {
+      ROW_SHL0 = 0x100,
+      ROW_SHR0 = 0x110,
+      ROW_ROR0 = 0x120,
+      WAVE_SHL1 = 0x130,
+      WAVE_ROL1 = 0x134,
+      WAVE_SHR1 = 0x138,
+      WAVE_ROR1 = 0x13C,
+      ROW_MIRROR = 0x140,
+      ROW_HALF_MIRROR = 0x141,
+      BCAST15 = 0x142,
+      BCAST31 = 0x143,
+    };
+
+    auto kind = DppOp.getKind();
+    auto permArgument = DppOp.getPermArgument();
+    int attrValue = 0;
+
+    switch (kind) {
+
+    case DPPPerm::quad_perm:
+      if (auto quadPermAttr = permArgument->dyn_cast<ArrayAttr>()) {
+        for (int i = 3; i >= 0; --i) {
+          auto intAttr = quadPermAttr[i].dyn_cast<IntegerAttr>();
+          int num = intAttr.getInt();
+          attrValue |= num << (i * 2);
+        }
+      }
+      break;
+    case DPPPerm::row_shl:
+      if (auto intAttr = permArgument->dyn_cast<IntegerAttr>()) {
+        attrValue = intAttr.getInt() + DppCtrl::ROW_SHL0;
+      }
+      break;
+    case DPPPerm::row_shr:
+      if (auto intAttr = permArgument->dyn_cast<IntegerAttr>()) {
+        attrValue = intAttr.getInt() + DppCtrl::ROW_SHR0;
+      }
+      break;
+    case DPPPerm::row_ror:
+      if (auto intAttr = permArgument->dyn_cast<IntegerAttr>()) {
+        attrValue = intAttr.getInt() + DppCtrl::ROW_ROR0;
+      }
+      break;
+    case DPPPerm::wave_shl:
+      attrValue = DppCtrl::WAVE_SHL1;
+      break;
+    case DPPPerm::wave_shr:
+      attrValue = DppCtrl::WAVE_SHR1;
+      break;
+    case DPPPerm::wave_rol:
+      attrValue = DppCtrl::WAVE_ROL1;
+      break;
+    case DPPPerm::wave_ror:
+      attrValue = DppCtrl::WAVE_ROR1;
+      break;
+    case DPPPerm::row_mirror:
+      attrValue = DppCtrl::ROW_MIRROR;
+      break;
+    case DPPPerm::row_half_mirror:
+      attrValue = DppCtrl::ROW_HALF_MIRROR;
+      break;
+    case DPPPerm::row_bcast_15:
+      attrValue = DppCtrl::BCAST15;
+      break;
+    case DPPPerm::row_bcast_31:
+      attrValue = DppCtrl::BCAST31;
+      break;
+    }
+
+    dppCtrl = rewriter.create<LLVM::ConstantOp>(loc, llvmI32Type, attrValue);
+
+    // Check for row_mask, bank_mask, bound_ctrl if they exist and create
+    // constants
+    auto rowMask =
+        createConstantOrUseDefault(DppOp, rewriter, "row_mask", llvmI32Type);
+    auto bankMask =
+        createConstantOrUseDefault(DppOp, rewriter, "bank_mask", llvmI32Type);
+    auto boundCtrl =
+        createConstantOrUseDefault(DppOp, rewriter, "bound_ctrl", llvmI1Type);
+
+    // create a ROCDL_DPPMovOp instruction with the appropriate attributes
+    auto dppMovOp = rewriter.create<ROCDL::DPPMovOp>(
+        loc, llvmI32Type, src, dppCtrl, rowMask, bankMask, boundCtrl);
+
+    Value result = dppMovOp.getRes();
+    if (srcType.getIntOrFloatBitWidth() < 32) {
+      result = rewriter.create<LLVM::TruncOp>(loc, llvmSrcIntType, result);
+    }
+
+    if (llvm::isa<FloatType>(srcType)) {
```

Comment:
To allow for more general usage, let's make this "not an integer type", as opposed to "is a float type". That way, you can `dpp.mov` a `vector<4xi8>`